rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,383 @@
+            size, p_engine_, g_alloc_);
+    // 2. Compute the value of c_buffer
+    // 3. Fulfill the promise
+    c_promise.set_value(c_buffer);
+}
+// ...
+// Execution end
+```
+
+Calling `get_or_add` in every execution may introduce overhead (such as hash
+table lookup, threads synchronization, ...). We need to reduce that overhead if
+it becomes model level performance bottleneck. Some options can be considered:
+
+- option 1: Further optimize the current `get_or_add` api implementation, like
+  using multi-level cache, more lightweight threads synchronization, ... To
+  reduce the function calling overhead.
+  - Pros: The interface is keep as simple as possible, and all backend could
+    have unified usage pattern.
+  - Cons: Performance improvement may be limited, since we need to keep the
+    interface general.
+- option 2: Extend the interfaces to allowing different programming model and
+  give backend more flexibilities to optimize cache performance by themselves.
+  - Pros: More flexible for each backend, more optimization opportunities and
+    performance improvement may be more significant.
+  - Cons: The interface becomes more complex, and the cache usage pattern in
+    different backend may be diverse, the backend API needs more time to
+    converge.
+
+We usually think it's more critical to optimize the overhead when cache hit.
+
+A recommendation is to add a `callback/notification` mechanism in unified cache,
+so that backend user can further hold the queried constant tensor handle by
+itself. When the tensor is still cached in unified cache, backend could directly
+access the handle efficiently without querying the unified cache every time.
+When the tensor is evicted from the unified cache, the callback should be called
+from unified cache side to tell backend the tensor is evicted, and backend will
+stop using the tensor. This propose could be treated as a specific case of above
+option 2.
+
+Backend API and integration optimization should not break the frontend API,
+unless there are new requests from framework users.
+
+### Multiple devices
+oneDNN Graph API allows to use multiple devices in two ways:
+- create engine with an `index` which specify the device index to use.
+- create engine with `sycl::device` and `sycl::context`.
+
+For multiple devices scenarios, there may be a situation that one device has
+been OOM since cached too much constant tensor, while the whole cache size still
+does not reach the capacity. Based on this consideration, we may need dedicated
+cache for each device and the constant tensor cache capacity actually becomes
+"the capacity for each device".
+
+The frontend capacity control API may be extended to the following. The engine
+kind parameter is replaced to engine, which contains both engine kind and device
+information.
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine engine, size_t size, policy_t policy);
```

Comment:
This is confusing as it looks like we bind constant tensor cache to engine. But for each device, we can create several engines.

Reply:
I update this part:
- bind cache to each device or device index instead of each engine
- discuss the weight sharing rules (only cps that use same device and context are allowed to share cached weight)

please help to review again, thanks.